### ISSUE TITLE
Add `pauli_rep` properties for common static gates

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#6243]
### Feature details

The `pauli_rep` property allows to represent an operator as a `PauliSentence` instance. 
Currently, it is mostly implemented for `qml.X, qml.Y, qml.Z` and for operator math classes.
It would be great to have `pauli_rep` attributes for, e.g., Clifford gates (`Hadamard`, `S`, `CNOT`), but also for other ops.
This would simplify manipulation of `qml.pauli` objects.

This came up while writing [this demo](https://github.com/PennyLaneAI/qml/pull/1209), where we currently have to do 
```python
new_op = qml.CNOT([0, 1]) @ original_op @ qml.CNOT([0, 1])
new_op = qml.pauli_decompose(new_op.matrix())
```
i.e., we have to go via the dense matrix representation to obtain the new operator. With `qml.CNOT.pauli_rep` implemented, we could stay in the more efficient Pauli representation.

### Implementation

Basically trivial, as far as I can tell.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by ldi18** [comment]
Hi, I’d like to work on this issue as my first contribution.

**Comment by dwierichs** [comment]
Hi @ldi18,
That's great to hear!
Let me start an internal discussion to see which operators should be included in this, and which constraints might apply.
I'll get back to you asap.

**Comment by dwierichs** [comment]
Hey @ldi18 
we'd be happy for you to contribute this feature.
I think a nice goal could be to include a Pauli representation for all gates in `pennylane/ops/qubit/non_parametric_ops.py`, and if you'd like you could also tackle `pennylane/ops/qubit/parametric_ops_single_qubit.py`.
After discussing internally, it would be favourable to evaluate the Pauli representations lazily, i.e., only when they are needed and not when the `Operation` is created, in order to avoid overhead for unrelated workflows.
Let me know in case you require any help with this. :)
As this is your first contribution, it will probably be helpful to take a look at our [how-to-PR page](https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html).

**Comment by ldi18** [comment]
Hi @dwierichs , 
I implemented Pauli representations for all ops in `non_parametric_ops` and `parametric_ops_single_qubit` as for the Hadamard gate here: 
```python

@property
def _pauli_rep(self):
    if self._pauli_rep_cached is None:
        self._pauli_rep_cached = qml.pauli.PauliSentence(
            {
                qml.pauli.PauliWord({self.wires[0]: "X"}): INV_SQRT2,
                qml.pauli.PauliWord({self.wires[0]: "Z"}): INV_SQRT2,
            }
        )
    return self._pauli_rep_cached

def __init__(self, wires: WiresLike, id: Optional[str] = None):
    super().__init__(wires=wires, id=id)
    self._pauli_rep_cached = None
```

I defined the test
```python
def test(gate):
    dim = gate.matrix().shape[0]
    if dim == 2:
        id_str = 'I(0)'
    if dim == 4:
        id_str = 'I(0)@I(1)'
    gate_in_pauli_decomp = eval(str(gate.pauli_rep).replace('\n', ' ').replace('I', id_str)).matrix()
    return np.allclose(gate.matrix(), gate_in_pauli_decomp)

```

All ops pass the test, but only if I comment out the assignment in the operator constructor as shown below:

```python
class Operator(abc.ABC, meta class=ABCCaptureMeta):
...
...
    def __init__(
        self,
        *params: TensorLike,
        wires: Optional[WiresLike] = None,
        id: Optional[str] = None,
    ):
        # pylint: disable=too-many-branches

        self._name: str = self.__class__.__name__  #: str: name of the operator
        self._id: str = id
        
        # Can this be commented out?
        # self._pauli_rep: Optional[qml.pauli.PauliSentence] = (
        #     None  # Union[PauliSentence, None]: Representation of the operator as a pauli sentence, if applicable
        # )
...
...)
```

Would it cause any issues elsewhere when I just remove the part I commented out above? To check this I tried running the tests as described in https://docs.pennylane.ai/en/stable/development/guide/tests.html . Actually this returns 11 errors, even for the current version in the GitHub - without any changes from me. 
 
Is this a problem with my testing environment or just normal? I'd just create a pull request now so that you can have a look.

**Comment by dwierichs** [comment]
Hi @ldi18 
That sounds great!
So we do not want to have any breaking change with the Operator API here. What I would suggest is trying to modify (instead of remove) the logic in `Operator` such that it is more permissive and allows you to do what you are already doing with the newly added (nice and lazy :) ) method. Probably you want to overwrite the `property` `pauli_rep` rather than `_pauli_rep`, no? :)

**Comment by ldi18** [comment]
Hi @dwierichs,

Thanks for the hint! Using pauli_rep resolved the issue in the operator class. During testing, however, `test_parametric_ops:TestDecompositions:test_pc_phase_decomposition_broadcasted` resulted in a shape error. Using an outer product in `pauli_arithmetic.py` for tensor-like phases (as in my pull request) would fix this.

My changes pass all pytest checks. Pylint still shows a few errors/warnings related to the number of parameters, but many of these also appear in the developer version without any changes from me.

**Comment by dwierichs** [comment]
Hi @ldi18 ,
Nice, happy to hear this worked.
At this point it's probably best to open the pull request :)
Then we can chat concretely about more nichy test failures and CI checks like pylint.

**Comment by dwierichs** [comment]
I would consider this issue closed by the contribution by @ldi18 above. @albi3ro any opinions on closing this one?
